vllm/inputs.py

```diff
@@ -127,4 +127,4 @@ class TextTokensPrompt(TypedDict):
 class LLMInputs(TypedDict):
     prompt_token_ids: List[int]
     prompt: NotRequired[Optional[str]]
-    multi_modal_data: NotRequired[Optional["MultiModalData"]]
+    multi_modal_data: NotRequired[Optional[Dict[str, Any]]]
```

Comment:
Let's provide a more concrete schema for built-in modalities.

```py
class MultiModalDataSchema(TypedDict, total=False):
    image: PIL.Image.Image
    # Can add other modalities in the future

# Still, we want users to be able to define additional modalities
MultiModalDataDict = Union[MultiModalDataSchema, Dict[str, Any]]

class LLMInputs(TypedDict):
    prompt_token_ids: List[int]
    prompt: NotRequired[Optional[str]]
    multi_modal_data: NotRequired[Optional[MultiModalDataDict]]
```

Reply:
Hmmm, for `LLM.generate()` user API, I prefer to only support the dict fashion and not expose internal data structures. 

For LLMInputs, since it's entirely external facing, I think it makes sense to have it as Dict[str, EXTERNAL_MM_DATA_TYPE].

For sequence.py stuff, I will make an alias so that the type hinting shows Union[MultiModalData, Dict[str, EXTERNAL_MM_DATA_TYPE]]. 
I am also going through the whole thing to make sure that type hinting is correct.

Reply:
No internal data structures are exposed. The TypedDict is also a dictionary but with each key  corresponding to specific data types (like PIL image)

Reply:
Ah I see. Then should we just enforce it to be always `MultiModalDataBuiltins` (remove `Dict[str, Any]`)?

Reply:
`Dict[str, Any]` should still be there to allow extra keys (such as modalities that are defined by users rather than the core library)

Reply:
ah I got where the confusion is.  what do you mean by "core library"? Is it the same as the entire vLLM library?
I guess in my mind, when one (a developer) adds a new modality, the following needs to happen:
- add the model to vllm/model_executor/models/
- write something equivalent to the current image.py that extends the base class. 
- add those to the registry.
- modify MultiModalDataBuiltins to reflect the newly added modality, or a newly accepted data type given an existing modality. 

Adding code is unavoidable?

Reply:
> ah I got where the confusion is.  what do you mean by "core library"? Is it the same as the entire vLLM library?
> I guess in my mind, when one (a developer) adds a new modality, the following needs to happen:
> - add the model to vllm/model_executor/models/
> - write something equivalent to the current image.py that extends the base class. 
> - add those to the registry.
> - modify MultiModalDataBuiltins to reflect the newly added modality, or a newly accepted data type given an existing modality. 
> 
> Adding code is unavoidable? 
> 
> 
Yes, by core library I mean our vLLM library .

The current plan is to allow users to develop their own plugins and register them through the registry without having to modify the code inside vLLM.

Reply:
synced offline. Apparently vLLM has already supported register external models through https://github.com/vllm-project/vllm/pull/3871.